### ISSUE TITLE
PostCSS Configuration

### ISSUE THREAD

**Issue by davezuko** [preactjs/wmr#250]
Hey preactjs team! I was super excited to find this project, and wanted to start by thanking you for all of your effort.

I saw that Sass support was added [a few months back](https://github.com/preactjs/wmr/commit/d60ad81cadd0f5d168f6340662736e878438aad4#diff-58be628d86f2ea957b662507ca5f8ea7fd9fd1931a3cda3dab28c663ac7381f2) and was curious to hear about your plans for PostCSS-oriented workflows.

The preact/htm/etc. ecosystem is enticing in its minimalism. Staying more "vanilla" is a major reason why this project is so appealing, and is also why I've moved away from Sass and toward PostCSS. And, unfortunately, as much as I'd _love_ to eliminate preprocessing entirely, some plugins (e.g. [Tailwind](https://tailwindcss.com/)) are too useful to give up and that requires a build step, even in development. So I set out to do that with wmr.

What I'm hoping to accomplish:
* Customize the PostCSS configuration. Currently it's [hardwired only for minifcation](https://github.com/preactjs/wmr/blob/4478690ebf806cfacd07380da8ba4fd5f4ce5963/src/plugins/minify-css-plugin.js#L5). I'm content to manage the added dependencies myself, I just need to be able to hook into it. Preferably with the native PostCSS configuration file for integration with other tooling.
* Run PostCSS in development. [From what I can see](https://github.com/preactjs/wmr/blob/4478690ebf806cfacd07380da8ba4fd5f4ce5963/src/bundler.js#L120), PostCSS only runs in production, so even if I could configure it, it wouldn't run in development. Worth considering here is that you could consider only do this if the project has its own PostCSS configuration. If not, skip it as you do now for speed.

I realize I could write a custom plugin for this. However, I raised this issue was because I hoped to avoid running PostCSS _twice_ in production. Perhaps doing so is a reasonable sacrifice for now, but at the same time I'm not sure there's much harm in delegating to the project's PostCSS configuration _if it exists_.

I'm aware that requests like this can be sort of a slippery slope, but thought it was worth bringing up since PostCSS is already a dependency in this project. Thank you for your attention :).

**Comment by developit** [comment]
Hi @davezuko! Running PostCSS once is definitely in keeping with the spirit of this project. For development I do like the idea of using the existence of a PostCSS config to turn on processing. 

This is quite a large undertaking - are you wanting to tackle it, or would it be helpful if I came up with a prototype?

**Comment by katywings** [comment]
@developit About that `Run PostCSS in development. `

I took yesterday some time to try if I can add the rollup transpilation to the css middleware part. `NoRollup.transform` has a little sideeffect with css, the returned code is not plain css anymore, but a js module which exports the transpiled css 😅☝. I didn't find a way to turn this of in transform so I see two ways how we could go about  it:
1. We could add a js "proxy" to css, analog to the cssModules middleware part (probably the better way, but might have unforeseen tradeoffs)
2. We could "extract" the css out of the rollup transform output (regex, eval and the likes) 🙈

Do you see better alternatives, or which of my ways do you prefer?

**Comment by davezuko** [comment]
@developit Thanks for the speedy reply. I don't intend to place the burden on you, but I do think it would be helpful to at least see what sort of implementation you'd envision. I will without a doubt try to work on it myself, and if I make reasonable progress I'll submit a PR. Until then, if you're able to start it off with a prototype, I'd happily build off of that work.

I haven't dug through the project enough yet to see how difficult it would be in development, but imagine there's (significantly?) more to consider than production builds (e.g. HMR).

---

I do also have an unrelated question for you, @developit, if you don't mind. I searched around for documentation about this but couldn't find anything, and want to be respectful of your issue tracker (or, if general questions are allowed there let me know!). Why is JSX converted to htm string templates in production? I can somewhat understand this format for development, but in production it would seem to result in less-minifiable code than `h()` function calls.

If I should move this question to a better place, just let me know where and I'll be sure to do so to keep this issue focused. Perhaps the slew of questions I have would be good inspiration for documentation (again, happy to help here, once I have the answers :P).

**Comment by marvinhagemeister** [comment]
> Why is JSX converted to htm string templates in production? I

A string is way faster to parse for JavaScript engines than hundreds of function calls and little objects. We're essentially trading a tiny tiny bit of runtime overhead for quicker parse times. The benefit of using `htm` specifically, is that it detects static vnodes and reuses them automatically. That information is lost with the usual JSX transforms.

**Comment by davezuko** [comment]
Thank you very much, @marvinhagemeister.

**Comment by jorbuedo** [comment]
Hi! Is there any progress with this? I would love to try wmr with tailwind.

**Comment by smeijer** [comment]
> Hi! Is there any progress with this? I would love to try wmr with tailwind.

Give it a shot with `twind/shim`! I've been doing so lately, and that works nicely.

Basically, just add an `import 'twind/shim'` to your entry file, and start using tailwind classnames.

If you're interested in server-side rendering, check this example repo:

https://github.com/tw-in-js/example-wmr/tree/7e33313847226a3659ce0316b4dd5aaf7c7f39ed

**Comment by jorbuedo** [comment]
Thanks! Didn't know about twind. I've been skeptical of the CSS-in-JS (prefered tailwind over styled), but this lib looks cool and their arguments over vanilla tailwind convinced me to try it.
Besides there's the easy integration with wmr! Having that I don't care about additional support for postcss.

**Comment by Siilwyn** [comment]
@smeijer this is a bit off-topic but I'm looking into wmr with tailwind to keep the bundle side small however with a SSR hello world project twind causes the JS bundle to jump from 12Kb to 46Kb. Do you know if it's possible to output just CSS?

**Comment by rschristian** [comment]
@Siilwyn Yes, you can, though you do lose all access to variant/directive grouping, as that's done at runtime.

What you'll want to do probably is hydrate with `twind` in dev, but in prod, hydrate with `preact-iso`.

**Comment by liflovs** [comment]
@Siilwyn what have you chosen as result? Using twind-cli feels a bit inconvenient, but adding extra 30+kb for nothing also is not a good case

**Comment by Siilwyn** [comment]
> @Siilwyn what have you chosen as result? Using twind-cli feels a bit inconvenient, but adding extra 30+kb for nothing also is not a good case

Nothing yet really, I've been looking at using other solutions than wmr.
